azure-iot-hub-devicesdk/azure/iot/hub/devicesdk/transport/mqtt/mqtt_transport.py

```diff
@@ -203,6 +217,39 @@ def _on_provider_publish_complete(self, mid):
             del self._event_callback_map[mid]
             callback()
 
+    def _on_provider_subscribe_complete(self, mid):
+        """
+        Callback that is called by the provider when a subscribe operation is complete.
+        """
+        logger.info("_on_provider_subscribe_complete")
+        callback = self._subscribe_callback
+        if callback:
+            self._subscribe_callback = None
+            callback()
+
+    def _on_provider_message_received_callback(self, topic, payload):
+        """
+        Callback that is called by the provider when a message is received.
+        """
+        logger.info(topic)
+        message_received = Message(payload)
+        # TODO : Discuss everything in bytes , need to be changed, specially the topic
+        topic_str = str(topic, "utf-8")
+        topic_parts = topic_str.split("/")
+
+        if "inputs" in topic_str and len(topic_parts) > 6:
+            # module input
+            input_name = topic_parts[5]
+            message_received.input_name = input_name
+            self._extract_properties(topic_parts[6], message_received)
+            self.on_transport_input_message_received(input_name, message_received)
+        elif "devicebound" in topic_str:
+            # device C2D
+            self._extract_properties(topic_parts[4], message_received)
+            self.on_transport_c2d_message_received(message_received)
+        else:
+            pass  # is there any other case
```

Comment:
>pass  # is there any other case [](start = 12, length = 31)

There will be more later because twins and methods will both need to go through this path, but this is good for now.